### ISSUE TITLE
How to remote configurations + remote state + multiple modules, goes wrong

### ISSUE THREAD

**Issue by zot24** [gruntwork-io/terragrunt#174]
I have to following and I'm not sure what I'm doing wrong, I want to be able to run two commands:

```shell
terragrunt apply # on the network folder to create just the network
```

and/or

```shell
terragrunt apply-all # on the root folder `terraform` to create network, database and compute resources
```

My project structure looks like:

<img width="327" alt="screen shot 2017-04-22 at 13 18 46" src="https://cloud.githubusercontent.com/assets/678498/25304399/607b55f6-275e-11e7-8c20-d9109d78e9c6.png">

My `terraform.tfvars` files looks like:

**terraform/network/terraform.tfvars**
```hcl
terragrunt = {
    terraform {
        source = "../../modules//network"
    },
    include {
        path = "${find_in_parent_folders()}"
    }
}
```

**terraform/terraform.tfvars**
```hcl
terragrunt = {
    terraform {
        extra_arguments "custom_vars" {
            commands = [
                "plan",
                "apply",
                "destroy"
            ]
            arguments = [
                "-var-file=${get_tfvars_dir()}/../variables.tfvars"
            ]
        }
    },
    remote_state {
        backend = "s3"
        config {
            bucket  = "my-bucket"
            key     = "aws/myapp/${path_relative_to_include()}/terraform.tfstat"
            region  = "eu-west-1"
        }
    }
}
```
I'm looking forward to move those `modules` into their own repo and then be able to do a proper `source=git::git@...` and pull the source code from git however I can't manage to make it work, it looks like the tmp folder got created properly and the code from my modules it's there however when it comes to run the terraform command it can not find the variables and terraform ask me to introduce them, on the tmp folder there are not `variables.tfvars` file so I guess that's why it's asking me for them.

I reckon I have misunderstood something and mixing concepts in here hope someone can point me out in the right direction, ta

**Comment by brikis98** [comment]
I think this is a duplicate of #176, which is a bug. I'll try to fix it tomorrow.

**Comment by brikis98** [comment]
I believe this should be fixed by #177. Please give [v0.12.3](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.12.3) a try (the binaries should show up once the build is done in a few minutes) and let me know if it works for you.

**Comment by brikis98** [comment]
Ah, hold up, there is one more bug. See #178 for details. I should have that merged shortly.

**Comment by brikis98** [comment]
OK, I think [v0.12.4](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.12.4) should fix all the bugs related to this... Please test and let me know :)

**Comment by zot24** [comment]
Hi @brikis98!

Still doesn't work for me quite as I expect so far I have realise that if I want to add `extra_arguments` like for example to load a global file for my variables I need to add that block on the inner `terraform.tfvars` files inside my modules and it won't work on the root `terrraform.tfvars` like I was trying.

So changing to this solve my problem however not sure if that the expected behaviour as that `extra_arguments` would be something that I'l gonna need to copy over all my modules as it won't change.

**terraform/network/terraform.tfvars**
```hcl
terragrunt = {
    terraform {
        source = "../../modules//network"
        extra_arguments "custom_vars" {
            commands = [
                "plan",
                "apply",
                "destroy"
            ]
            arguments = [
                "-var-file=${get_tfvars_dir()}/../variables.tfvars"
            ]
        }
    },
    include {
        path = "${find_in_parent_folders()}"
    }
}
```

**terraform/terraform.tfvars**
```hcl
terragrunt = {
    remote_state {
        backend = "s3"
        config {
            bucket  = "my-bucket"
            key     = "aws/myapp/${path_relative_to_include()}/terraform.tfstat"
            region  = "eu-west-1"
        }
    }
}
```

That's one thing the other thing is it still doesn't create the remote state on S3 and instead it use the `terraform.tfstate` therefore when the `database` module that depends on `network` try to use the data `terraform_remote_state` it just fail.

It's weird because I can see how it initialize the backend

```
.../terraform/network] 2017/04/24 14:16:14 Running command: terraform init -backend-config=bucket=my-bucket -backend-config=key=aws/myapp/network/terraform.tfstat -backend-config=region=eu-west-1
```

And the other thing that happens and doesn't looks normal (not sure if related) is that looks like Terragrunt try to run all command on the root directory where there is nothing more than just the global `terraform.tfvars` 

```
2017/04/24 [/myapp/terraform] 14:16:09 Running command: terraform init -backend-config=bucket=my-bucket -backend-config=key=aws/myapp/./terraform.tfstat -backend-config=region=eu-west-1
Terraform initialized in an empty directory!

The directory has no Terraform configuration files. You may begin working
with Terraform immediately by creating Terraform configuration files.
[terragrunt] [/myapp/terraform] 2017/04/24 14:16:09 Running command: terraform apply -input=false
No configuration files found!

Apply requires configuration to be present. Applying without a configuration
would mark everything for destruction, which is normally not what is desired.
If you would like to destroy everything, please run 'terraform destroy' instead
which does not require any configuration files.
```

**Comment by brikis98** [comment]
> Still doesn't work for me quite as I expect so far I have realise that if I want to add extra_arguments like for example to load a global file for my variables I need to add that block on the inner terraform.tfvars files inside my modules and it won't work on the root terrraform.tfvars like I was trying.

Not sure I follow. Is the issue that the relative paths in your `extra_arguments` isn't working since the path is different for each child module? If so, check out the [helper functions](https://github.com/gruntwork-io/terragrunt#helper-functions), especially `path_relative_from_include`, `get_tfvars_dir`, and `get_parent_tfvars_dir` that were all added in recent releases.

> That's one thing the other thing is it still doesn't create the remote state on S3 and instead it use the terraform.tfstate therefore when the database module that depends on network try to use the data terraform_remote_state it just fail.

Did you define a `backend` in your Terraform configurations (`.tf` files)?

```hcl
terraform {
  # You must define this in your Terraform code or Terragrunt won't be able to configure it
  backend "s3" {}
}
```

**Comment by zot24** [comment]
Was just about to write you that I'm missing the `backend` on my `.tf` gonna give it another go and report back.

Thanks for your help

**Comment by zot24** [comment]
The S3 problem looks like is solve 👍 

> Not sure I follow. Is the issue that the relative paths in your extra_arguments isn't working since the path is different for each child module? If so, check out the helper functions, especially path_relative_from_include, get_tfvars_dir, and get_parent_tfvars_dir that were all added in recent releases.

I think the problem with the `extra_arguments` it's not related with the `helper functions` when it work I can see that the path is correct.

I think the problem is when merging the `terraform` block from the root with the child `terraform.tfvars`. When `extra_arguments` is on the root `terraform.tfvars` it doesn't work however if I move the `extra_arguments` to the `terraform.tfvars` on the child nodes it works properly.

That's how my `extra_arguments` looks like:

```
extra_arguments "custom_vars" {
        commands = [
            "plan",
            "apply",
            "destroy"
        ]
        arguments = [
            "-var-file=${get_tfvars_dir()}/../variables.tfvars",
            "-var-file=terraform.tfvars"
        ]
    }
```

Still can't get the `terragrunt apply-all` to work I'll keep trying...

**Comment by zot24** [comment]
Is there any reason why `FindConfigFilesInPath` it's returning even the `terraform.tfvars` on the root folder and not just the subfolders on that directory? https://github.com/gruntwork-io/terragrunt/blob/4dda205db9b5cda99ce80d56394a667b503da1d7/config/config.go#L100 that's causing me that every time I run `terraform apply-all` from the root folder it trying to run `terraform init` and `terraform apply` to the root folder where I have my global `terraform.tfvars` and therefore see the following errors:

```
[terragrunt] [X/terraform] 2017/04/24 17:38:37 Running command: terraform init -backend-config=bucket=my-bucket -backend-config=key=aws/myapp/./terraform.tfstat -backend-config=region=eu-west-1
Terraform initialized in an empty directory!

The directory has no Terraform configuration files. You may begin working
with Terraform immediately by creating Terraform configuration files.
[terragrunt] [X/terraform] 2017/04/24 17:38:37 Running command: terraform apply -input=false
No configuration files found!

Apply requires configuration to be present. Applying without a configuration
would mark everything for destruction, which is normally not what is desired.
If you would like to destroy everything, please run 'terraform destroy' instead
which does not require any configuration files.
```

Note: Something that's was happening as well it's that the tmp dir get into a dirty state coz I'm testing different options and configurations so I need to wipe it out before continue testing that I think was causing me some weird errors and unexpected behaviour.

**Comment by zot24** [comment]
> I think the problem with the extra_arguments it's not related with the helper functions when it work I can see that the path is correct.

Ok, one thing less, I have realise what's the problem with the `extra_arguments` the whole block `terraform` on the included config it's being override by the current config, in my case the module config which container the `source` code to pull the module code:

**terraform/network/terraform.tfvars**
```hcl
terragrunt = {
    terraform {
        source = "../../modules//network"
    },
    include {
        path = "${find_in_parent_folders()}"
    }
}
```

**terraform/terraform.tfvars**
```hcl
terragrunt = {
    terraform {
        extra_arguments "custom_vars" {
            commands = [
                "plan",
                "apply",
                "destroy"
            ]
            arguments = [
                "-var-file=${get_tfvars_dir()}/../variables.tfvars"
            ]
        }
    },
    remote_state {
        backend = "s3"
        config {
            bucket  = "my-bucket"
            key     = "aws/myapp/${path_relative_to_include()}/terraform.tfstat"
            region  = "eu-west-1"
        }
    }
}
```

I though it will just override the values of the variables that collide inside the `terraform` block and not the whole block like mentioned [here](https://github.com/gruntwork-io/terragrunt/blob/4dda205db9b5cda99ce80d56394a667b503da1d7/config/config.go#L249):

```
// Merge the given config with an included config. Anything specified in the current config will override the contents
// of the included config. If the included config is nil, just return the current config.
func mergeConfigWithIncludedConfig(config *TerragruntConfig, includedConfig *TerragruntConfig) (*TerragruntConfig, error) {
...
```

**Comment by zot24** [comment]
Sorry I'm piling up thing in this closed issue I just want to be sure of what I'm doing and if there is a real bug that will deserve a new issue or it just me being clumsy 😃 and I hope this could help someone at some point

> Is there any reason why FindConfigFilesInPath it's returning even the terraform.tfvars on the root folder and not just the subfolders on that directory?

I have realise that this my be just how Terragrunt works and is an expected behaviour it collide with my folder structure as I'm not using any subfolder to wrap the environment of the project (prod, dev, qa) so I was running `terragrunt apply-all` from `terraform` folder on the following structure:

<img width="327" alt="screen shot 2017-04-22 at 13 18 46" src="https://cloud.githubusercontent.com/assets/678498/25304399/607b55f6-275e-11e7-8c20-d9109d78e9c6.png">

Which was causing the problem mentioned before I end doing some modifications and now my folder structure looks like:

![screen shot 2017-04-24 at 21 56 37](https://cloud.githubusercontent.com/assets/678498/25358333/f4228a3c-2938-11e7-8dcb-e1a36e2af234.png)

And I get no errors if I run Terragrunt from `terrafrom/production` folder as there is no `terrafrom.tfvars` on that folder

**Comment by brikis98** [comment]
> I think the problem is when merging the terraform block from the root with the child terraform.tfvars. When extra_arguments is on the root terraform.tfvars it doesn't work however if I move the extra_arguments to the terraform.tfvars on the child nodes it works properly.

Yes, Terragrunt currently does a very crude merge, just gluing the top-level keys together. So if you specify `remote_state` in the parent and `remote_state` in the child, the latter completely overrides the former. The problem with a more granular merge is that there is no way to "unset" a key. If the parent specifies `remote_state.foo = "bar"`, with a granular merge, there is no way for a child to override that key to something that isn't set at all (note: in Terraform "not set" is sometimes not treated the same as "set to empty string").

I don't know the best solution to this problem. I proposed an idea here that may be worth exploring: https://github.com/gruntwork-io/terragrunt/issues/147#issuecomment-282276693

> I have realise that this my be just how Terragrunt works and is an expected behaviour it collide with my folder structure as I'm not using any subfolder to wrap the environment of the project (prod, dev, qa) so I was running terragrunt apply-all from terraform folder on the following structure:

Yes, your new folder structure looks correct.
